### ISSUE TITLE
Improve vmmap on Darwin

### ISSUE THREAD

**Issue by patryk4815** [pwndbg/pwndbg#2565]
Currently, on Darwin systems, the vmmap command often labels many pages as "unknown."

## 1. parse dyld_shared_cache
It would be helpful to identify memory mappings originating from the dyld_shared_cache and replace the "unknown" label with "dyld_shared_cache" where applicable.

![image](https://github.com/user-attachments/assets/5bb1cda0-fb69-46d9-9be5-c4dfa7b6f82d)

Additionally, it might be worth introducing a command like `vmmap --show-dyld-cache` to expand those large 500 MB memory allocations into smaller mappings. Since the cache contains around 2,900 libraries, it’s probably unnecessary to display all of them by default. Such detailed output might be excessive for most use cases.

Example file:
```
# dyld_shared_cache_arm64e.map
# dyld_shared_cache_x86_64h.map

$ cat /System/Volumes/Preboot/Cryptexes/OS/System/Library/dyld/dyld_shared_cache_x86_64h.map

mapping  EX  583MB 0x7FF800000000 -> 0x7FF82472C000
mapping  RW   28MB 0x7FF840000000 -> 0x7FF841C14000
mapping  RW   31MB 0x7FF841C14000 -> 0x7FF843B50000
mapping  RO  180MB 0x7FF880000000 -> 0x7FF88B418000
mapping  EX  512MB 0x7FF900000000 -> 0x7FF920044000
mapping  RW   29MB 0x7FF940000000 -> 0x7FF941D88000
mapping  RW   41MB 0x7FF941D88000 -> 0x7FF9446B0000
mapping  RO  194MB 0x7FF980000000 -> 0x7FF98C24C000
mapping  EX  517MB 0x7FFA00000000 -> 0x7FFA20564000
mapping  RW   30MB 0x7FFA40000000 -> 0x7FFA41EF8000
mapping  RW   33MB 0x7FFA41EF8000 -> 0x7FFA44094000
mapping  RO  133MB 0x7FFA80000000 -> 0x7FFA8857C000
mapping  EX  510MB 0x7FFB00000000 -> 0x7FFB1FEFC000
mapping  RW   12MB 0x7FFB40000000 -> 0x7FFB40CF4000
mapping  RW   45MB 0x7FFB40CF4000 -> 0x7FFB43A9C000
mapping  RO  120MB 0x7FFB80000000 -> 0x7FFB878D0000
mapping  EX  511MB 0x7FFC00000000 -> 0x7FFC1FF4C000
mapping  RW    8MB 0x7FFC40000000 -> 0x7FFC4085C000
mapping  RW   54MB 0x7FFC4085C000 -> 0x7FFC43E6C000
mapping  RO  152MB 0x7FFC80000000 -> 0x7FFC898A0000
mapping  EX   87MB 0x7FFD00000000 -> 0x7FFD0572C000
mapping  RW    1MB 0x7FFD40000000 -> 0x7FFD4010C000
mapping  RW   10MB 0x7FFD4010C000 -> 0x7FFD40BD4000
mapping  RO  176MB 0x7FFD80000000 -> 0x7FFD8B0FC000


/usr/lib/libobjc.A.dylib
                  __TEXT 0x7FF800068000 -> 0x7FF8000A9FC4
            __DATA_CONST 0x7FF840000000 -> 0x7FF840001C48
                  __DATA 0x7FF841C14000 -> 0x7FF841C18694
            __DATA_DIRTY 0x7FF84397CCC0 -> 0x7FF84397FB50
              __LINKEDIT 0x7FF880000000 -> 0x7FF88B418000

/usr/lib/dyld
                  __TEXT 0x7FF8000AA000 -> 0x7FF80013A820
            __DATA_CONST 0x7FF840001C50 -> 0x7FF840007310
                  __DATA 0x7FF841C186C0 -> 0x7FF841C1B380
            __DATA_DIRTY 0x7FF84397FB80 -> 0x7FF8439816E0
              __LINKEDIT 0x7FF880000000 -> 0x7FF88B418000

/usr/lib/system/libsystem_blocks.dylib
                  __TEXT 0x7FF80013B000 -> 0x7FF80013EFF9
            __DATA_CONST 0x7FF840007310 -> 0x7FF840007850
                  __DATA 0x7FF841C1B380 -> 0x7FF841C1B3B0
            __DATA_DIRTY 0x7FF8439816E0 -> 0x7FF8439818D8
              __LINKEDIT 0x7FF880000000 -> 0x7FF88B418000
```

## 2. parse `type:` information 
 - `debugserver` only expose small set of memory types:  *(eg. missing type: share memory)*
 https://github.com/llvm/llvm-project/blob/c894d3a846f351e592bf917ace57c7d2fcbeddc8/lldb/tools/debugserver/source/MacOSX/MachVMRegion.cpp#L186-L225
 - xnu Darwin kernel has a lot of memory types:
 https://github.com/apple-oss-distributions/xnu/blob/8d741a5de7ff4191bf97d57b9f54c2f6d4a15585/osfmk/mach/vm_statistics.h#L581-L785
- lldb can read `type:` only heap/stack:
https://github.com/llvm/llvm-project/blob/a87f776c1c873ea86a95c368f1a9331adc65d1ee/lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp#L1638-L1645
 - another issue is that LLDB don't expose "raw" `type:` via python api:
![image](https://github.com/user-attachments/assets/dbb9c79c-709c-4fa9-afe9-47dcb73789e8)
- program `vmmap <pid>` prints more into eg: `REGION DETAIL` + `REGION TYPE`:
https://github.com/apple/darwin-xnu/blob/main/osfmk/mach/vm_region.h
<img width="1087" alt="image" src="https://github.com/user-attachments/assets/0e20f78e-0a1f-4711-adca-164bced10fd2">

### Debuging lldb on macOS:
```
# in lldb
log enable gdb-remote all

# in lldb (debugging debugserver)
settings set target.process.extra-startup-command QSetLogging:bitmask=LOG_ALL;

# in another console:
log stream --process debugserver --style compact
```
https://reverse.put.as/2019/11/19/how-to-make-lldb-a-real-debugger/

**Comment by patryk4815** [comment]
Apple COMMPAGE
> This page is mapped into every userland process and contains various variables exposed by the kernel to userland.


<img width="308" alt="Image" src="https://github.com/user-attachments/assets/6badc976-de0f-4a6e-88a4-121e519efd90" />

https://github.com/apple-oss-distributions/xnu/blob/main/osfmk/arm/commpage/commpage.c#L98
https://blog.svenpeter.dev/posts/m1_sprr_gxf/

**Comment by mbrla0** [comment]
Looks like the output of `vmmap` got fixed at some point? I've just finished writing enumeration functionality for images in the shared cache, and when I went to compare the output of both to make sure I was getting the right base addresses, I noticed it was already fixed.

<img width="1232" height="1349" alt="Image" src="https://github.com/user-attachments/assets/e99ea2c2-379e-4002-b3a8-8d93f557e66f" />

**Comment by patryk4815** [comment]
@mbrla0 I think it is still invalid. In vmmap we see 1 big ~4gb page, but "inside" are a lot of libraries

**Comment by mbrla0** [comment]
> [@mbrla0](https://github.com/mbrla0) I think it is still invalid. In vmmap we see 1 big ~4gb page, but "inside" are a lot of libraries

Yeah, fair enough, the sizes are all wonky still, and I didn't see that.

**Comment by mbrla0** [comment]
Well, good to know my work enumerating the libs wasn't wasted.

**Comment by patryk4815** [comment]
Probably we should "rename" current logic with string: `[dyld_shared_cache]`, just to not confuse people
Right now we display a lot of random crap:

<img width="383" height="239" alt="Image" src="https://github.com/user-attachments/assets/abfcf3f8-92c3-4f46-ac9f-557a1cec2ac9" />

**Comment by patryk4815** [comment]
reported upstream issue:
- https://github.com/llvm/llvm-project/issues/155692
- https://github.com/llvm/llvm-project/issues/155945
